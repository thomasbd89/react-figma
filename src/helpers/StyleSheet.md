# StyleSheet

A StyleSheet is an abstraction similar to CSS StyleSheets.
Analog of the `StyleSheet` module at [React Native](https://facebook.github.io/react-native/docs/stylesheet), 
[React Sketchapp](http://airbnb.io/react-sketchapp/docs/API.html#stylesheet).

## API

`create(styles)`

Create an optimized StyleSheet reference from the style object.

`flatten(styles)`

Flatten an array of style objects into one aggregated object, or look up the definition for a registered stylesheet.

`resolve(style)`

Resolve one style.

`compose(style1, style2)`

Combines two styles such that `style2` will override any styles in `style1`.

`hairlineWidth`

This is defined as the width of a thin line on the platform. Returns 1.

## Example

```typescript
const styles = StyleSheet.create({
  foo: {
    fontSize: 24,
    color: 'red',
  },
  bar: {
    fontSize: 36,
    color: 'blue',
  },
});
// { foo: 1, bar: 2 }

<View>
  <Text style={styles.foo} />
  <Text style={styles.bar} />
</View>;
```
