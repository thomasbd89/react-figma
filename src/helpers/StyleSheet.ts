import { CommonStyle } from '../types';

type StyleProp<P, T> = { [P in keyof T]: Partial<CommonStyle> };

export class StyleSheet {
    static create<T extends StyleProp<any, T>>(styles: T): T {
        return styles;
    }

    static flatten<C extends Partial<CommonStyle>>(styles: C | C[] | void): C {
        if (Array.isArray(styles)) {
            return styles.reduce((acc, item) => ({ ...acc, ...item }), {} as any);
        } else {
            return styles || ({} as C);
        }
    }

    static resolve<T extends { [key: string]: Partial<CommonStyle> }>(value: T): T {
        return value;
    }

    static compose<C1 extends StyleProp<any, C1>, C2 extends StyleProp<any, C2>>(style1: C1, style2: C2): C1 & C2 {
        const result = { ...style1, ...style2 } as (C1 & C2);
        Object.keys(style1).map(style1Key => {
            if (style2[style1Key]) {
                result[style1Key] = { ...style1[style1Key], ...style2[style1Key] };
            }
        });
        return result;
    }
}
