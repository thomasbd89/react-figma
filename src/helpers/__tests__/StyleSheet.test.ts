import { StyleSheet } from '../StyleSheet';

describe('StyleSheet', () => {
    it('compose', () => {
        const style1 = StyleSheet.create({
            container: {
                width: '100%'
            },
            inner: {
                backgroundColor: 'green',
                margin: 20
            },
            text: {
                fontFamily: 'Helvetica',
                fontSize: 24
            }
        });

        const style2 = StyleSheet.create({
            container: {
                height: '100%'
            },
            inner: {
                backgroundColor: 'red'
            },
            icon: {
                width: 20,
                height: 20
            }
        });

        const result = StyleSheet.compose(
            style1,
            style2
        );
        expect(result).toMatchSnapshot();
    });
});
