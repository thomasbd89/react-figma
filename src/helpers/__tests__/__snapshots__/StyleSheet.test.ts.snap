// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyleSheet compose 1`] = `
Object {
  "container": Object {
    "height": "100%",
    "width": "100%",
  },
  "icon": Object {
    "height": 20,
    "width": 20,
  },
  "inner": Object {
    "backgroundColor": "red",
    "margin": 20,
  },
  "text": Object {
    "fontFamily": "Helvetica",
    "fontSize": 24,
  },
}
`;
